### DISCUSSION TITLE
The Best Way to Add TCA To A Large Application That Already Exists

### DISCUSSION THREAD

**Discussion by brentbrinkley** [pointfreeco/swift-composable-architecture#1165] [Q&A] [answered]
I work for a company that already has a large production application in place. Our state is kind of a mess and I want to fix it. What's the best way to adopt this slowly into a project like this? I would like to start at a feature and slowly work back until the entire codebase is converted over. Is that even possible with the way TCA is designed?

**Comment by tgrapperon** [comment, marked as answer]
Hey @brentbrinkley!
This is definitely what I would suggest. I find easier to implement features backward, because you know what the child domain looks like. It is a little harder to advance forward in the dark, especially if you're starting with TCA. Each TCA feature can be seen as an autonomous app that can work in complete isolation. By building backward, you're recursively inheriting all the "good" traits of your children, and generally mostly have to handle what's different at this level. As a consequence, if you find an issue, you generally have to fix it only in one place. Furthermore, working backward makes it harder to leak top-level features into leaf domains, which is generally what happens with apps that went out of control.
